Search filter has been implemented

LakshyaBagani opened the pull request:
Fixes **#16749**

## **Summary**

Adds **search keywords** to `ManagementLink` items on the **Manage Jenkins** page, allowing users to find configuration pages by searching for terms like **“proxy”**, **“csrf”**, or **“kubernetes”**, instead of matching only display names.

## **Problem**

Previously, when using the **Search Settings** box on the `/manage` page:

* Search results matched **only display names** such as *Plugins* or *Security*
* Searching for actual configuration concepts (e.g. **proxy**, **usage statistics**, **CSRF**) returned **no results**
* Users had to manually browse categories despite the settings being present

This reduced discoverability and made navigation unintuitive, especially for new users.

## **Solution**

This change enhances the search behavior by:

* Adding **relevant search keywords** to `ManagementLink` items
* Allowing searches to match **both display names and common configuration terms**
* Preserving all existing search behavior

## **Testing Done**

### **Search Functionality**

* Built Jenkins locally and started the development server
* Navigated to **Manage Jenkins** page → search box displayed correctly
* Verified keyword-based search results:

  * Searching **“proxy”** → *Plugins* link appears
  * Searching **“csrf”** → *Security* link appears
  * Searching **“kubernetes”** → *Clouds* link appears
  * Searching **“logging”** → *System Log* link appears
* Verified original behavior still works:

  * Searching **“Plugins”** still finds *Plugins*

### **Edge Cases Tested**

* Verified empty keywords do not break search
* Confirmed search is **case-insensitive**
* Tested **partial keyword matching**
  (e.g. **“kube”** matches **“kubernetes”**)

## **Proposed Changelog Entry**

* Add search keywords to the **Manage Jenkins** page for improved settings discoverability

## **Proposed Changelog Category**

/label rfe

## **Upgrade Guidelines**

N/A

After the change
<img width="1839" height="918" alt="image" src="https://github.com/user-attachments/assets/62c23d0c-51dc-4e6d-8f21-c6fed89e0c34" />

Before the change
<img width="1850" height="926" alt="Screenshot from 2025-12-13 01-47-46" src="https://github.com/user-attachments/assets/3b152c3a-13aa-4c5b-830d-868836f8b00f" />


## **Submitter Checklist**

* [x] The issue is well described
* [x] Changelog entry is appropriate for affected users
* [x] Testing is documented
* [x] New public APIs are properly annotated or avoided
* [x] No UI regressions introduced under current CSP rules
* [x] No dependency updates involved

## **Desired Reviewers**

@timja @daniel-beck

LakshyaBagani committed:
earch filter has been implemented

LakshyaBagani committed:
Fix spotless: remove unused imports

LakshyaBagani committed:
Fix the issue

daniel-beck reviewed (changes requested):
While I agree that the search on Manage Jenkins needs to be better (since moving the form field around, it feels less of a basic filter and more like an actual search), this approach isn't how to get there.

As minimum, this needs to account for Jenkins's extensibility. This may have a reasonable result with the naked core feature set, but there are multiple expected results (General, Tools, Credentials) when typing `git` in practically every realistic setup.

> Searching “csrf” → Security link appears

This is an interesting example, as I anticipate the removal of that option from most instances in https://github.com/jenkinsci/jenkins/pull/25918. Some random other code location would also need to be adapted with this PR.

LakshyaBagani commented:
Thanks for the feedback. To address your concerns:
    1. Extensibility: Make getSearchKeywords() automatically use displayName and description as fallback, so plugins work 
         without overriding.       
    2 . Multiple results: Increase limit to 10 and add relevance ranking (exact matches first).
    3 . Stale keywords: Defaulting to description/displayName keeps keywords in sync; explicit overrides can be documented 
         for review.
Does this approach work? Should I proceed with implementation?

LakshyaBagani committed:
Search issue fixe

LakshyaBagani committed:
Fix raw type warning and add missing @Override annotation

- Changed Descriptor to Descriptor<?> to avoid raw type warnings
- Added @Override annotation to getIconFileName() method

LakshyaBagani committed:
Fix Prettier formatting in .vscode/settings.json

LakshyaBagani committed:
Delete .vscode/settings.json

LakshyaBagani commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -2431,6 +2431,12 @@ protected Iterable<TopLevelItem> allAsIterable() {
+
+        // Add GlobalConfiguration items to search index so users can find configuration sections
+        for (Descriptor<?> d : Functions.getSortedDescriptorsForGlobalConfigUnclassifiedReadable()) {
+            builder.add("/manage/configure", d.getDisplayName());
```
You're right on both points.
The ConfigureLink.getSearchKeywords() already adds the GlobalConfiguration descriptors for the Manage Jenkins page filter. The code in Jenkins.java was adding the same items to the global search index, which is a separate system and expands the scope beyond the original fix.
I'll remove the changes from Jenkins.java.

LakshyaBagani replied:
Hi @daniel-beck and @mawinter69,
Thanks again for the earlier feedback — I’ve addressed the points raised:

* Removed outdated keywords (e.g., proxy from Plugins)
* Added GlobalConfiguration display names to Configure System search keywords
* Removed unrelated global search index changes
* Cleaned up formatting and unnecessary files

At this point, the PR is focused solely on improving the **Manage Jenkins page search discoverability** without affecting other search systems.

Could you please let me know if any further changes are needed, or if this is ready for another review?
Appreciate your time and guidance!

LakshyaBagani committed:
Remove unrelated global search index changes from Jenkins.java

The ConfigureLink.getSearchKeywords() already handles the "Search Settings"
box on the Manage Jenkins page. The code in Jenkins.java was adding items
to the global Jenkins search (header search box), which is a different
system and unrelated to the search settings fix.

LakshyaBagani commented:
@MarkEWaite @mawinter69 @timja @daniel-beck 

Just following up on this PR. Please let me know if any changes or improvements are needed from my side.
